### PR TITLE
FileSystems should throw exception if filesystem for schema is not found

### PR DESCRIPTION
**Pull request by angoenka**
Not failing and falling back to LocalFileSystem creates hard to find bug and delayed failure.

Example:
File path: gs://my_bucker/my_folder
Running the program from /root/
If GCS is not added then in the current code we create a local file "/root/gs://my_bucker/my_folder" without any exception or message.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Spark
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by angoenka** [comment]
R: @lukecwik @jkff

**Comment by jkff** [comment]
Thanks - I've hit this issue many times too, but will this work correctly on Windows? (for paths like c:/path/to/file.txt)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 1]
```diff
@@ -470,10 +470,7 @@ static FileSystem getFileSystemInternal(String scheme) {
     String lowerCaseScheme = scheme.toLowerCase();
     Map<String, FileSystem> schemeToFileSystem = SCHEME_TO_FILESYSTEM.get();
     FileSystem rval = schemeToFileSystem.get(lowerCaseScheme);
-    if (rval != null) {
-      return rval;
-    }
-    return schemeToFileSystem.get(DEFAULT_SCHEME);
+    return checkNotNull(rval);
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: I wouldn't use a `checkNotNull` but an explicit `IllegalArgumentException` saying no handler found for scheme `foo`.

Comment (angoenka) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: Will Fix this.

**Comment by angoenka** [comment]
No, It will not work for windows. The patter is 
`<schema>:*` which will extract "c" as schema in this case.
We can change the pattern to 
`<schema>://*` to fix this.

**Comment by jkff** [comment]
Sounds reasonable. Or check that scheme has more than 1 character.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java [hunk 2]
```diff
@@ -68,6 +68,11 @@ public void testGetLocalFileSystem() throws Exception {
     assertTrue(
         FileSystems.getFileSystemInternal(toLocalResourceId("File://home").getScheme())
             instanceof LocalFileSystem);
+    if (SystemUtils.IS_OS_WINDOWS) {
+      assertTrue(
+          FileSystems.getFileSystemInternal(toLocalResourceId("c:\\home\\").getScheme())
+              instanceof LocalFileSystem);
+    }
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java]: Can you also add a test for "No filesystem found for scheme "?

Comment (angoenka) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java]: thanks @jkff . Added the test cases

**Comment by jkff** [comment]
Thanks!

**Pull request merged**